attestation/link/link.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2024 The Witness Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package link
+
+import (
+	"encoding/json"
+	"fmt"
+
+	v0 "github.com/in-toto/attestation/go/predicates/link/v0"
+	v1 "github.com/in-toto/attestation/go/v1"
+	"github.com/in-toto/go-witness/attestation"
+	"github.com/in-toto/go-witness/attestation/commandrun"
+	"github.com/in-toto/go-witness/attestation/environment"
+	"github.com/in-toto/go-witness/attestation/material"
+	"github.com/in-toto/go-witness/attestation/product"
+	"github.com/in-toto/go-witness/cryptoutil"
+	"github.com/in-toto/go-witness/registry"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+const (
+	Name    = "link"
+	Type    = "https://in-toto.io/attestation/link/v0.3"
+	RunType = attestation.PostProductRunType
+
+	defaultExport = false
+)
+
+// This is a hacky way to create a compile time error in case the attestor
+// doesn't implement the expected interfaces.
+var (
+	_ attestation.Attestor  = &Link{}
+	_ attestation.Subjecter = &Link{}
+)
+
+func init() {
+	attestation.RegisterAttestation(Name, Type, RunType,
+		func() attestation.Attestor { return New() },
+		registry.BoolConfigOption(
+			"export",
+			"Export the link attestation to its own file",
+			defaultExport,
+			func(a attestation.Attestor, export bool) (attestation.Attestor, error) {
+				linkAttestor, ok := a.(*Link)
+				if !ok {
+					return a, fmt.Errorf("unexpected attestor type: %T is not a link attestor", a)
+				}
+				WithExport(export)(linkAttestor)
+				return linkAttestor, nil
+			},
+		),
+	)
+}
+
+type Option func(*Link)
+
+func WithExport(export bool) Option {
+	return func(l *Link) {
+		l.export = export
+	}
+}
+
+type Link struct {
+	PbLink   v0.Link
+	products map[string]attestation.Product
+	export   bool
+}
+
+func New() *Link {
+	return &Link{}
+}
+
+func (l *Link) Name() string {
+	return Name
+}
+
+func (l *Link) Type() string {
+	return Type
+}
+
+func (l *Link) RunType() attestation.RunType {
+	return RunType
+}
+
+func (l *Link) Export() bool {
+	return l.export
+}
+
+func (l *Link) Attest(ctx *attestation.AttestationContext) error {
+	l.PbLink.Name = "stepNameHere"
+	for _, attestor := range ctx.CompletedAttestors() {
```

Comment:
So this is less a comment specific to this code, but more something i'm wondering more generally:

> Should an attestor be able to inspect the metadata generated by other attestors? Even if the attestor is native to the tool (not a custom attestor like I have proposed elsehwere), is it _safe_ to attestors to basically interact with each others data?

Reply:
https://github.com/in-toto/go-witness/pull/149/#discussion_r1478435889

Reply:
Was this meant to another PR? this is just taking me to the top of this PR 😃 .

Reply:
We need to inspect to output from other attestors.  For example in the OCI attestor we query the product attestor for candidates.  It is the reason why we needed to create the lifecycle